Add tuple add operator

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52292 Add tuple add operator**

Fixes https://github.com/pytorch/pytorch/issues/51967, https://github.com/pytorch/pytorch/issues/55249, https://github.com/pytorch/pytorch/issues/55250

Differential Revision: [D26792416](https://our.internmc.facebook.com/intern/diff/D26792416)

tugsbayasgalan committed:
temp

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "temp"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "temp"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "temp"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add tuple add operator"


Fixes https://github.com/pytorch/pytorch/issues/51967

Differential Revision: [D26792416](https://our.internmc.facebook.com/intern/diff/D26792416)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/runtime/register_ops_utils.cpp:
```diff
@@ -369,6 +370,31 @@ void listAdd(Stack* stack) {
+
+  for (const auto& el : a->elements()) {
+    // don't add None to tuple
+    if (!el.isNone()) {
```
Why skipping `None`s? I tried following python code, which seems to not skip `None`
```
>>> a = (1, None, 2)
>>> b = (3, None, 4)
>>> a + b
(1, None, 2, 3, None, 4)
```

gmagogsfm commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -69,18 +70,23 @@ IValue toIValue(py::handle obj, const TypePtr& type, c10::optional<int32_t> N) {
       size_t tuple_size = tuple.size();
       auto tuple_type = type->cast<TupleType>();
       const auto& elem_types = tuple_type->elements();
-      if (elem_types.size() != tuple_size) {
```
Why is this check removed?

tugsbayasgalan replied:
This is because annotation like Tuple[Tensor, Optional[Tensor]] would think the tuple size is 2 while python thinks it is size of 1 (e.g (torch.rand(4),). My idea was instead of checking size here, we just need to make sure that we are inserting None.

gmagogsfm commented on test/test_jit.py:
```diff
@@ -11188,6 +11188,23 @@ def return_tuple(x):
             return a, x
         self.checkScript(return_tuple, (torch.rand(4),))
 
+    def test_add_tuple_slice(self):
```
Consider adding a test case for adding two tuples (without slicing).

gmagogsfm commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -69,18 +70,23 @@ IValue toIValue(py::handle obj, const TypePtr& type, c10::optional<int32_t> N) {
+
+      for (size_t i = tuple_size; i < elem_types.size(); i++) {
+        if (elem_types[i]->kind() != TypeKind::OptionalType) {
+          throw py::cast_error(c10::str(
```
Could you elaborate what this limitation is about? what would be an example in Python that triggers this error?

tugsbayasgalan committed:
Update on "Add tuple add operator"


Fixes https://github.com/pytorch/pytorch/issues/51967

Differential Revision: [D26792416](https://our.internmc.facebook.com/intern/diff/D26792416)

[ghstack-poisoned]

Lilyjjo commented on test/test_jit.py:
```diff
@@ -11188,6 +11188,21 @@ def return_tuple(x):
+            changed_input = input[0] + 1
+            return (changed_input,) + input[1:]
+        inp: Tuple[torch.Tensor, Optional[torch.Tensor], Optional[torch.Tensor]] = (torch.rand(4), None, None)
+        self.checkScript(foo, (inp,))
```
Yay! This will make the users of the hooks much happier until we can get the type inferred hooks PR written

lobstergrindset reviewed (approved):
This looks good to me. You might want to get another person to approve, I'm not super familiar with this part of the code base, but the logic seems clean !

eellison commented on test/test_jit.py:
```diff
@@ -11188,6 +11188,21 @@ def return_tuple(x):
+    def test_add_tuple_slice_non_optional(self):
+        def foo(input: Tuple[torch.Tensor, torch.Tensor, torch.Tensor]):
+            changed_input = input[0] + 1
+            return (changed_input,) + input[1:]
```
Can you add an example where the return type of the tuple add is used inside of a ScriptFunction ? `AnyTupleType` refers to `tuple` which is not type that you can do anything with inside of script function, it's kind of just an abstract supertype of all tuples. 

I think you're going to need to type the return value more accurately than `AnyTupleType` to be useable in TorchScript

gmagogsfm replied:
+1

Current test cases don't actually verify the output type is computed correctly. The case suggested by Elias would cover it.

Or, even better you can use FileCheck to verify the type is correct in generated IR

tugsbayasgalan replied:
I replaced them with return statements like value[3] or torch.sum(value[3]).

tugsbayasgalan committed:
Update on "Add tuple add operator"


Fixes https://github.com/pytorch/pytorch/issues/51967

Differential Revision: [D26792416](https://our.internmc.facebook.com/intern/diff/D26792416)

[ghstack-poisoned]

eellison commented:
It would be nice as a follow up or as part of this PR to add lowering of adding tuples to [lower tuples](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/lower_tuples.cpp). All of our existing tuple operators can be decomposed and lowered which is important for exporting to other backends and optimization.

tugsbayasgalan committed:
Update on "Add tuple add operator"


Fixes https://github.com/pytorch/pytorch/issues/51967

Differential Revision: [D26792416](https://our.internmc.facebook.com/intern/diff/D26792416)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add tuple add operator"


Fixes https://github.com/pytorch/pytorch/issues/51967

Differential Revision: [D26792416](https://our.internmc.facebook.com/intern/diff/D26792416)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add tuple add operator"


Fixes https://github.com/pytorch/pytorch/issues/51967, https://github.com/pytorch/pytorch/issues/55249, https://github.com/pytorch/pytorch/issues/55250

Differential Revision: [D26792416](https://our.internmc.facebook.com/intern/diff/D26792416)

[ghstack-poisoned]

tugsbayasgalan commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3484,6 +3484,25 @@ struct to_ir {
+      }
+      for (auto i = 0; i < second_tuple.size(); i++) {
+        tuple_values.push_back(second_tuple.at(i));
+      }
```
done

tugsbayasgalan committed:
Update on "Add tuple add operator"


Fixes https://github.com/pytorch/pytorch/issues/51967, https://github.com/pytorch/pytorch/issues/55249, https://github.com/pytorch/pytorch/issues/55250

Differential Revision: [D26792416](https://our.internmc.facebook.com/intern/diff/D26792416)

[ghstack-poisoned]
